waymo_raw_data_to_argoverse.py

```diff
@@ -30,205 +40,398 @@
+                    frame,
+                    range_images,
+                    camera_projections,
+                    range_image_top_pose,
+                    ri_index=1,
+                )
+            points_all_ri = np.concatenate(points_ri, axis=0)
+            if args.save_cloud:
+                dump_point_cloud(points_all_ri, timestamp_ns, log_id, ARGO_WRITE_DIR)
+            # Saving labels
+            if args.save_labels:
+                dump_object_labels(
+                    frame.laser_labels,
+                    timestamp_ns,
+                    log_id,
+                    ARGO_WRITE_DIR,
+                    track_id_dict,
+                )
+            if args.save_calibration:
+                calib_json = form_calibration_json(frame.context.camera_calibrations)
+                if log_calib_json is None:
+                    log_calib_json = calib_json
+                    calib_json_fpath = (
+                        f"{ARGO_WRITE_DIR}/{log_id}/vehicle_calibration_info.json"
+                    )
+                    check_mkdir(str(Path(calib_json_fpath).parent))
+                    save_json_dict(calib_json_fpath, calib_json)
+                else:
+                    assert calib_json == log_calib_json
+
+            # 5 images per frame
+            for index, tf_cam_image in enumerate(frame.images):
+                # 4x4 row major transform matrix that transforms
+                # 3d points from one frame to another.
+                SE3_flattened = np.array(tf_cam_image.pose.transform)
+                city_SE3_egovehicle = SE3_flattened.reshape(4, 4)
+                # in seconds
+                timestamp_s = tf_cam_image.pose_timestamp
+                timestamp_ns = int(timestamp_s * 1e9)  # to nanoseconds
+                if args.save_poses:
+                    dump_pose(city_SE3_egovehicle, timestamp_ns, log_id, ARGO_WRITE_DIR)
+
+                if args.save_images:
+                    camera_name = CAMERA_NAMES[tf_cam_image.name]
+                    img = tf.image.decode_jpeg(tf_cam_image.image)
+                    new_img = undistort_image(
+                        np.asarray(img),
+                        frame.context.camera_calibrations,
+                        tf_cam_image.name,
+                    )
+                    img_save_fpath = f"{ARGO_WRITE_DIR}/{log_id}/{camera_name}/{camera_name}_{timestamp_ns}.jpg"
+                    # assert not Path(img_save_fpath).exists()
+                    check_mkdir(str(Path(img_save_fpath).parent))
+                    imageio.imwrite(img_save_fpath, new_img)
+                    img_count += 1
+                    if img_count % 100 == 0:
+                        print(f"\tSaved {img_count}'th image for log = {log_id}")
+
+
+def undistort_image(img: np.ndarray, calib_data: Any, camera_name: int):
+    """Undistort the image from the Waymo dataset given camera calibration data"""
```

Comment:
do you mind briefly mentioning what k1, k2, k3 and p1, p2 represent here? (tangential vs. radial)?

Reply:
can we add a brief unit test here as well, suppose we start with a 40x40 white image with a black diagonal from bottom-left to top-right, and top-left to bottom-right? THen we distort it ourselves, and make sure we can recover a clean "X" when we call this function?